builtin_set.go

```diff
@@ -344,3 +344,90 @@ func (r *Runtime) getSet() *Object {
 	}
 	return ret
 }
+
+// estimateMemUsage helps calculating mem usage for large objects.
+// It will sample the object and use those samples to estimate the
+// mem usage.
+func (so *setObject) estimateMemUsage(ctx *MemUsageContext) (estimate uint64, err error) {
+	var samplesVisited, memUsage uint64
+	totalItems := so.m.size
+	if totalItems == 0 {
+		return memUsage, nil
+	}
+	sampleSize := ctx.ComputeSampleStep(totalItems)
+
+	// We can use samplesVisited instead of an index since we iterate using
+	// iterNext
+	for item := so.m.iterFirst; item != nil && samplesVisited < uint64(totalItems); item = item.iterNext {
+		if samplesVisited%uint64(sampleSize) != 0 {
+			continue
+		}
+		samplesVisited += 1
+		inc, err := item.key.MemUsage(ctx)
+		memUsage += inc
+		if err != nil {
```

Comment:
[q] if there is an error, can `inc` for the key be a non-zero value? If so, do we to also calculate the value as well?

Reply:
Not too sure I follow but we do `memUsage += inc` right before the error which should account for any non-zero increment.

Reply:
so the `item.key.MemUsage` may error and my thought is that inc can still have a value (or maybe it never does on error?) I am assuming an `item.key` has a corresponding `item.value`, so if the item.key errors out with a value for inc, does that mean we are not catching the inc value for `item.value`? Maybe I am overthinking it

Reply:
Aaah yes, that makes sense. I think it's a very rare case but IF it were to happen it might skew the estimate a bunch since key/value are tied together. I'll make the change since it's pretty trivial, nice catch!